src/observer/consume.py

```diff
@@ -0,0 +1,49 @@
+import backoff, requests, requests_cache
+from datetime import timedelta
+from django.conf import settings
+import logging
+
+LOG = logging.getLogger(__name__)
+
+logging.getLogger('backoff').setLevel(logging.CRITICAL)
+
+if settings.DEBUG:
+    requests_cache.install_cache(**{
+        'cache_name': '/tmp/api-cache',
+        'backend': 'sqlite',
+        'fast_save': True,
+        'extension': '.sqlite3',
+        # https://requests-cache.readthedocs.io/en/latest/user_guide.html#expiration
+        'expire_after': timedelta(hours=24)
+    })
+
+def _giving_up(details):
+    LOG.error("request %s failed after %s attempts", details['args'][0], details['tries'])
+
+def _retrying(details):
+    LOG.warn("re-attempting request: %s", details['args'][0])
+
+@backoff.on_exception(
```

Comment:
Hadn't seen `backoff` before, but had written some of its functionality multiple times in various projects. Will consider it next time. 👍

Reply:
exactly. faced with having to write yet-another-exp-backoff function I just gave it a quick google and voila. it's docs aren't great, but the readme is thorough and those decorators are flexible in what they accept